src/main/java/events/writer/PubSubEventWriter.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package events.writer;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.grpc.InstantiatingGrpcChannelProvider;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.gson.Gson;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.pubsub.v1.TopicName;
+import events.Event;
+import io.grpc.HttpConnectProxiedSocketAddress;
+import io.grpc.ProxiedSocketAddress;
+import io.grpc.ProxyDetector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.*;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+
+/**
+ * PubSubEventWriter takes messages from CDAP Event Publisher and writes them
+ * to Google Cloud PubSub
+ */
+public class PubSubEventWriter implements EventWriter {
+    private Publisher publisher;
+    private static final Logger logger = LoggerFactory.getLogger(PubSubEventWriter.class);
+
+    /**
+     *
+     * @param projectId
+     * @param serviceAccountPath
+     * @param topicId
+     * @param proxyHost
+     * @param proxyPort
+     */
+    public PubSubEventWriter(String projectId, String serviceAccountPath, String topicId, String proxyHost, String proxyPort) {
```

Comment:
The constructor should be a zero-argument one since this class will be loaded as an extension. All the properties required can be read using a context passed to an initialize() method and the code for creating publisher can be moved there. (The extension loading is based on Java ServiceLoader - https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html ) .

Reply:
Use metrics-exporter-extension as example to build the Context object to pass the different instead the full constructor

Reply:
Repo link here for another extension - https://github.com/cdapio/metrics-writer-extension

Reply:
done